packages/flutter/lib/src/widgets/lazy_block.dart

```diff
@@ -0,0 +1,449 @@
+      _minScrollOffset = currentLogicalOffset - blockExtent * 2.0;
+      _startOffsetLowerLimit = currentLogicalOffset;
+    }
+
+    // Materialize new children until we fill the viewport (or run out of
+    // children to materialize).
+
+    RenderBox child;
+    while (currentLogicalOffset < endLogicalOffset) {
+      if (!_materializeChild(builder, currentLogicalIndex))
+        break;
+      child = _getNextWithin(block, child);
+      assert(child != null);
+      if (currentLogicalIndex >= firstLogicalIndexNeedingLayout) {
+        assert(currentLogicalIndex == firstLogicalIndexNeedingLayout);
+        child.layout(innerConstraints, parentUsesSize: true);
+        firstLogicalIndexNeedingLayout += 1;
+      }
+      currentLogicalOffset += _getMainAxisExtent(child.size);
+      currentLogicalIndex += 1;
+    }
+
+    // We now have all the physical children we ought to have to fill the
+    // viewport. The currentLogicalIndex is the index of the first child that
+    // we don't need.
+
+    if (currentLogicalOffset < endLogicalOffset) {
+      // The last element is visible. We need to update our reckoning of where
+      // the max scroll offset is.
+      _maxScrollOffset = currentLogicalOffset;
+      _startOffsetUpperLimit = double.INFINITY;
+    } else {
+      // The last element is not visible. Ensure that we have enough headroom
+      // so we don't hit the max scroll offset prematurely.
+      _maxScrollOffset = currentLogicalOffset + blockExtent * 2.0;
+      _startOffsetUpperLimit = currentLogicalOffset - blockExtent;
+    }
+
+    // Remove any unneeded children.
+
+    int currentPhysicalIndex = currentLogicalIndex - _firstChildLogicalIndex;
+    final int numberOfRequiredPhysicalChildren = currentPhysicalIndex;
+    while (currentPhysicalIndex < _children.length) {
+      deactivateChild(_children[currentPhysicalIndex]);
+      _children[currentPhysicalIndex] = null;
+      currentPhysicalIndex += 1;
+    }
+    _children.length = numberOfRequiredPhysicalChildren;
+
+    // We now have the correct physical children, each of which has gone through
+    // layout exactly once. We still need to position them correctly. For
+    // simplicity, we position the first physical child at Offset.zero and use
+    // the paintOffset on the render object to adjust the final paint location
+    // of the children.
+
+    Offset currentChildOffset = Offset.zero;
+    child = block.firstChild;
+    while (child != null) {
+      final _LazyBlockParentData childParentData = child.parentData;
+      childParentData.offset = currentChildOffset;
+      currentChildOffset += _getMainAxisOffsetForSize(child.size);
+      child = childParentData.nextSibling;
+    }
+
+    _updatePaintOffset();
+
+    LazyBlockExtentsChangedCallback onExtentsChanged = widget.onExtentsChanged;
+    if (onExtentsChanged != null)
+      onExtentsChanged(_maxScrollOffset - _minScrollOffset, blockExtent, _minScrollOffset);
```

Comment:
iirc you should only call this if the extents actually changed.

Also, it'd be best if we called this from outside the layout callback scope, since otherwise the callback can fiddle with the child list with impunity.